### PR TITLE
[feat] Add PR review summary comment

### PR DESCRIPTION
**Pull request by rohitvinnakota-codecov**
Example: https://github.com/codecov/codecov-api/pull/1285#issuecomment-2813280012

Screenshot 2025-04-23 at 9 56 50 AM
Screenshot 2025-04-23 at 9 56 55 AM

This is v0 that lays the groundwork for a more robust PR summary comment. We want to merge overwatch, coverage, TA, and any issue context here in the future

### CODE REVIEW

File: src/seer/automation/codegen/pr_review_coding_component.py [hunk 1]
```diff
@@ -45,6 +45,19 @@ def invoke(
...
+                messages=agent.memory,
+                prompt=CodingCodeReviewPrompts.format_pr_description_step(
+                    diff_str=request.diff,
+                ),
+                model=AnthropicProvider.model("claude-3-5-sonnet-v2@20241022"),
+                run_name="Generate PR description",
+                max_tokens=2048,
+            )
+
+            print(pr_description.message.content, "YO CHECK ME OUT")
```
Comment (giovanni-guidini) [inline src/seer/automation/codegen/pr_review_coding_component.py]: [nit] i know it's still a draft PR, but print

File: src/seer/automation/codegen/prompts.py [hunk 2]
```diff
@@ -223,6 +223,30 @@ def pr_review_formatter_msg():
...
+    @staticmethod
+    def format_pr_description_step(diff_str: str):
+        return textwrap.dedent(
+            """\
+            You are given the below code changes as a diff:
+            {diff_str}
+
+            # Your goal:
+            Provide a detailed, high-level description of what this pull request is doing. Focus on:
+            - The main purpose and goal of the changes
```
Comment (giovanni-guidini) [inline src/seer/automation/codegen/prompts.py:235]: From your example it seems the AI breaks up these into sections. It looks cool, but it's probably quite verbose. I'd love that _before_ I open the PR, not after I took the time to write it myself.

Is this feature gonna be available before opening PRs? If so the only remark I'd make is to set some limits on the section length

Otherwise maybe intruct the LLM to "create at most 3 paragraphs for the entire analysis" or something like that, so it's not too verbose?

**Review by giovanni-guidini — APPROVED** [review]
Very nice example, sounds like a cool feature 👍

File: src/seer/automation/codegen/prompts.py (hunk 2 above)
Comment (rohitvinnakota-codecov) [reply src/seer/automation/codegen/prompts.py:235]: At the moment, this will be triggered if the user comments `@ ai reviewer review` OR if they have auto PR reviews enabled. 

Auto PR reviews only run on PRs that are opened or ready for review in a non draft state. My thinking behind this was PRs in a draft state are usually not ready for review anyway, and the developer is more likely to push changes at that point making the review stale. 

I think as a follow up, I will look into structured outputs, and additional data can go into a toggle dropdown so it doesn't take up as much room.

File: src/seer/automation/codegen/models.py [hunk 3]
```diff
@@ -194,7 +194,16 @@ class Comment(BaseModel):
         start_line: int
         suggestion: str | None = None
 
+    class PrDescription(BaseModel):
+        purpose: str
+        key_technical_changes: str
+        architecture_decisions: str
+        dependencies_and_interactions: str
+        risk_considerations: str
+        notable_implementation_details: str
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/models.py:203]: The `PrDescription` model has good structure but could benefit from field validation to ensure non-empty strings and reasonable length limits. Consider adding Pydantic field validators with `constr` type for string length constraints and non-empty validation.
```suggestion
class PrDescription(BaseModel):
    purpose: constr(min_length=1, max_length=2000)
    key_technical_changes: constr(min_length=1, max_length=2000)
    architecture_decisions: constr(min_length=1, max_length=2000)
    dependencies_and_interactions: constr(min_length=1, max_length=2000)
    risk_considerations: constr(min_length=1, max_length=2000)
    notable_implementation_details: constr(min_length=1, max_length=2000)
```

File: src/seer/automation/codegen/pr_review_coding_component.py [hunk 4]
```diff
@@ -45,6 +45,18 @@ def invoke(
...
+            # Get detailed PR description
+            pr_description = llm_client.generate_structured(
+                messages=agent.memory,
+                prompt=CodingCodeReviewPrompts.format_pr_description_step(
+                    diff_str=request.diff,
+                ),
+                model=GeminiProvider(model_name="gemini-2.0-flash-001"),
+                response_format=CodePrReviewOutput.PrDescription,
+                run_name="Generate PR description",
+                max_tokens=2048,
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/pr_review_coding_component.py:57]: The max_tokens=2048 limit for PR description generation might be too restrictive for complex PRs. Consider making this configurable or increasing it to ensure comprehensive descriptions. Also, consider adding error handling for when description generation fails.
```suggestion
max_tokens=4096  # Increased token limit for more detailed descriptions
```

File: src/seer/automation/codegen/pr_review_publisher.py [hunk 5]
```diff
@@ -29,6 +29,25 @@ def publish_generated_pr_review(self, pr_review: CodePrReviewOutput) -> None:
...
+                    f"### Key Technical Changes\n{pr_review.description.key_technical_changes}\n\n"
+                    f"### Architecture Decisions\n{pr_review.description.architecture_decisions}\n\n"
+                    f"### Dependencies and Interactions\n{pr_review.description.dependencies_and_interactions}\n\n"
+                    f"### Risk Considerations\n{pr_review.description.risk_considerations}\n\n"
+                    f"### Notable Implementation Details\n{pr_review.description.notable_implementation_details}\n"
+                    f"</details>"
+                )
+                repo_client.post_issue_comment(pr_url, description_content)
+            except ValueError as e:
+                logger.warning(f"Failed to post PR description on PR {pr_url}: {e}")
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/pr_review_publisher.py:49]: The description posting logic could be moved to a separate method for better code organization and reusability. Also, consider adding retries for the GitHub API calls as they can sometimes be flaky.
```suggestion
def _post_pr_description(self, description: CodePrReviewOutput.PrDescription) -> None:
    try:
        description_content = (
            f"## PR Description\n\n"
            f"{description.purpose}\n\n"
            f"<details>\n"
            f"<summary><b>Click to see more</b></summary>\n\n"
            f"### Key Technical Changes\n{description.key_technical_changes}\n\n"
            f"### Architecture Decisions\n{description.architecture_decisions}\n\n"
            f"### Dependencies and Interactions\n{description.dependencies_and_interactions}\n\n"
            f"### Risk Considerations\n{description.risk_considerations}\n\n"
            f"### Notable Implementation Details\n{description.notable_implementation_details}\n"
            f"</details>"
        )
        for attempt in range(3):  # Add retries
            try:
                self.repo_client.post_issue_comment(self.pr.url, description_content)
                break
            except ValueError as e:
                if attempt == 2:  # Last attempt
                    raise
                time.sleep(1 * (attempt + 1))  # Exponential backoff
    except ValueError as e:
        logger.warning(f"Failed to post PR description on PR {self.pr.url}: {e}")
```

File: tests/automation/codegen/test_pr_review_coding_component.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        self.assertEqual(
+            formatted_response_call[1]["response_format"], list[CodePrReviewOutput.Comment]
+        )
+        self.assertEqual(formatted_response_call[1]["run_name"], "Generate PR review structured")
+        self.assertEqual(formatted_response_call[1]["max_tokens"], 8192)
+
+        # Verify suggestion was added to comment body
+        self.assertIn(f"\n```suggestion\n{mock_comment.suggestion}\n```", mock_comment.body)
+
+        # Verify result
```
Comment (rohitvinnakota-codecov) [inline tests/automation/codegen/test_pr_review_coding_component.py:144]: The tests are thorough but could benefit from testing edge cases where the LLM returns malformed or unexpected data. Consider adding test cases for when PR description fields contain empty strings or extremely long content.
```suggestion
@patch("seer.automation.codegen.pr_review_coding_component.LlmClient")
def test_invoke_handles_malformed_pr_description(self, mock_llm_client_class):
    # Mock PR description with empty/invalid fields
    mock_pr_description = MagicMock()
    mock_pr_description.parsed = MagicMock(spec=CodePrReviewOutput.PrDescription)
    mock_pr_description.parsed.purpose = ""
    mock_pr_description.parsed.key_technical_changes = "" * 3000  # Too long
    
    mock_formatted_response = MagicMock()
    mock_formatted_response.parsed = []
    
    mock_llm_client_class.generate_structured.side_effect = [
        mock_pr_description,
        mock_formatted_response,
    ]
    
    # Should handle invalid data gracefully
    result = self.component.invoke(self.request, mock_llm_client_class)
    self.assertIsNotNone(result)
    self.assertIsNone(result.description)  # Invalid description should be ignored
```

File: src/seer/automation/codegen/prompts.py [hunk 7]
```diff
@@ -223,6 +223,30 @@ def pr_review_formatter_msg():
...
+            - Dependencies or interactions with other parts of the system
+            - Any potential risks or considerations
+
+            # Guidelines:
+            - Be concise but comprehensive
+            - Use clear, technical language
+            - Focus on the "why" as much as the "what"
+            - Highlight any significant changes that reviewers should pay special attention to
+            - If there are multiple distinct changes, organize them logically
+            """
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/prompts.py:247]: The PR description prompt could benefit from examples to guide the LLM better. Also, consider adding specific guidance about security implications and performance considerations in the prompt.
```suggestion
def format_pr_description_step(diff_str: str):
    return textwrap.dedent(
        """\
        You are given the below code changes as a diff:
        {diff_str}

        # Your goal:
        Provide a detailed, high-level description of what this pull request is doing. Focus on:
        - The main purpose and goal of the changes
        - Key technical changes and their impact
        - Any notable architectural decisions or patterns
        - Dependencies or interactions with other parts of the system
        - Any potential risks or considerations
        - Security implications of the changes
        - Performance impact considerations

        # Guidelines:
        - Be concise but comprehensive
        - Use clear, technical language
        - Focus on the "why" as much as the "what"
        - Highlight any significant changes that reviewers should pay special attention to
        - If there are multiple distinct changes, organize them logically
        - Explicitly mention any security or performance implications
        
        Example structure:
        Purpose: "This PR implements rate limiting for the API endpoints to prevent abuse"
        Key Technical Changes: "Added Redis-based rate limiting middleware, configured limits per endpoint"
        ...
        """
    ).format(diff_str=diff_str)
```

File: src/seer/automation/codegen/models.py (hunk 3 above)
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/models.py:203]: Consider adding docstrings to the PrDescription model class and its fields to provide better documentation of the expected content and format for each field. This will help developers understand what kind of information should go into each field.

File: src/seer/automation/codegen/models.py [hunk 8]
```diff
@@ -194,7 +194,16 @@ class Comment(BaseModel):
         start_line: int
         suggestion: str | None = None
 
+    class PrDescription(BaseModel):
+        purpose: str
+        key_technical_changes: str
+        architecture_decisions: str
+        dependencies_and_interactions: str
+        risk_considerations: str
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/models.py:202]: Consider adding field validators to ensure non-empty strings and reasonable length limits for the PR description fields. This will help maintain quality of the generated descriptions and prevent potential issues with very long content.
```suggestion
class PrDescription(BaseModel):
    purpose: str = Field(..., min_length=1, max_length=2000)
    key_technical_changes: str = Field(..., min_length=1, max_length=2000)
    architecture_decisions: str = Field(..., min_length=1, max_length=2000)
    dependencies_and_interactions: str = Field(..., min_length=1, max_length=2000)
    risk_considerations: str = Field(..., min_length=1, max_length=2000)
    notable_implementation_details: str = Field(..., min_length=1, max_length=2000)
```

File: src/seer/automation/codegen/pr_review_coding_component.py [hunk 9]
```diff
@@ -45,6 +45,18 @@ def invoke(
...
+            )
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/pr_review_coding_component.py:58]: Consider adding a timeout parameter or circuit breaker for the LLM calls. Long-running LLM requests could impact system performance. Also, consider adding error handling for LLM failures to make the system more robust.

File: src/seer/automation/codegen/pr_review_publisher.py [hunk 10]
```diff
@@ -29,6 +29,25 @@ def publish_generated_pr_review(self, pr_review: CodePrReviewOutput) -> None:
...
+                    f"<summary><b>Click to see more</b></summary>\n\n"
+                    f"### Key Technical Changes\n{pr_review.description.key_technical_changes}\n\n"
+                    f"### Architecture Decisions\n{pr_review.description.architecture_decisions}\n\n"
+                    f"### Dependencies and Interactions\n{pr_review.description.dependencies_and_interactions}\n\n"
+                    f"### Risk Considerations\n{pr_review.description.risk_considerations}\n\n"
+                    f"### Notable Implementation Details\n{pr_review.description.notable_implementation_details}\n"
+                    f"</details>"
+                )
+                repo_client.post_issue_comment(pr_url, description_content)
+            except ValueError as e:
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/pr_review_publisher.py:48]: The error handling in the PR description publishing could be more specific. Instead of catching a general ValueError, consider catching specific exceptions that might occur during the API call and providing more detailed error logging.
```suggestion
try:
    repo_client.post_issue_comment(pr_url, description_content)
except (GithubException, APIError) as e:
    logger.warning(f"Failed to post PR description on PR {pr_url}. Error type: {type(e).__name__}, Details: {str(e)}")
```

File: src/seer/automation/codegen/pr_review_publisher.py (hunk 5 above)
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/pr_review_publisher.py:49]: Consider adding retry logic for the PR description publishing in case of temporary GitHub API issues. This would make the system more resilient to transient failures.

File: tests/automation/codegen/test_pr_review_publisher.py [hunk 11]
```diff
@@ -136,3 +140,56 @@ def test_format_comments_filtering(self):
...
+                f"<summary><b>Click to see more</b></summary>\n\n"
+                f"### Key Technical Changes\n{mock_pr_description.key_technical_changes}\n\n"
+                f"### Architecture Decisions\n{mock_pr_description.architecture_decisions}\n\n"
+                f"### Dependencies and Interactions\n{mock_pr_description.dependencies_and_interactions}\n\n"
+                f"### Risk Considerations\n{mock_pr_description.risk_considerations}\n\n"
+                f"### Notable Implementation Details\n{mock_pr_description.notable_implementation_details}\n"
+                f"</details>"
+            )
+
+            mock_repo_client.post_issue_comment.assert_called_with(
```
Comment (rohitvinnakota-codecov) [inline tests/automation/codegen/test_pr_review_publisher.py:193]: Consider adding test cases for different types of PR description content, including edge cases like very long descriptions or descriptions with special characters. This would help ensure robust handling of various content types.

File: src/seer/automation/codegen/prompts.py [hunk 12]
```diff
@@ -223,6 +223,30 @@ def pr_review_formatter_msg():
...
+        ).format(diff_str=diff_str)
```
Comment (rohitvinnakota-codecov) [inline src/seer/automation/codegen/prompts.py:248]: Consider adding some validation or sanity checks for the diff_str parameter in format_pr_description_step. Very large diffs might need special handling or truncation to work effectively with the LLM.

File: src/seer/automation/codegen/pr_review_coding_component.py (hunk 4 above)
Comment (jennmueng) [reply src/seer/automation/codegen/pr_review_coding_component.py:57]: this does seem a little bit restrictive, maybe double it to 4096? Makes sense that we want to keep it concise, but we can do that via the prompt, because if it hits the token limit here it'll just error out.

File: src/seer/automation/codegen/prompts.py [hunk 13]
```diff
@@ -223,6 +223,30 @@ def pr_review_formatter_msg():
             Format the comments into a list of JSON objects, ready to be used in a GitHub pull request review. Please ensure you insert a comma between each JSON object except for the last one."""
         )
 
+    @staticmethod
+    def format_pr_description_step(diff_str: str):
+        return textwrap.dedent(
+            """\
+            You are given the below code changes as a diff:
+            {diff_str}
```
Comment (jennmueng) [inline src/seer/automation/codegen/prompts.py:231]: as a v0 this is fine, but I think it would be useful to do a more thorough context gathering step in order to provide a more "in context" description of the PR where it better understands internal terminology and how the system works together as a whole, that would probably prove highly useful to a reviewer.

**Review by jennmueng — APPROVED** [review]

File: src/seer/automation/codegen/prompts.py (hunk 13 above)
Comment (rohitvinnakota-codecov) [reply src/seer/automation/codegen/prompts.py:231]: For sure, thanks
